streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManager.java

```
@@ -194,12 +204,18 @@ public KafkaFuture<Void> lockTasks(final Set<TaskId> taskIds) {
                 }
 
                 if (assignedTasks.containsKey(taskId)) {
-                    final KafkaFuture<StreamTask> future = assignedTasks.get(taskId).unassign();
+                    final TaskExecutor executor = assignedTasks.get(taskId);
+                    log.debug("Requesting release of task {} from {}", taskId, executor.name());
+                    final KafkaFuture<StreamTask> future = executor.unassign();
                     future.whenComplete((streamTask, throwable) -> {
                         if (throwable != null) {
                             result.completeExceptionally(throwable);
                         } else {
-                            remainingTaskIds.remove(streamTask.id());
+                            assert !assignedTasks.containsKey(taskId);
+                            // It can happen that the executor handed back the task before we asked it to
+                            // in which case `streamTask` will be null here.
+                            assert streamTask == null || streamTask.id() == taskId;
```

Usually, we do not use `assert` but rather `IllegalStateException`. Is there a good reason to use `assert`?

---

You use assertions to add extra checks to fail early during development, refactoring and testing which simplifies debugging and also serves as executable documentation. However, assertions will be disabled in production runs, both to not slow down the code and to avoid crashing in production just because the assertion is wrong.

Using `IllegalStateException` in these places can backfire. An example is this bug, which caused several incidents: https://github.com/apache/kafka/pull/13534 where a "this should not happen here" kind of check regressed into not being true anymore.

`IllegalStateException` may still make sense if you want to avoid something from happening that is worse than crashing, or if you want to catch it.